ucp/_libs/ucp_py.pyx

```diff
@@ -729,7 +729,18 @@ def get_endpoint(peer_ip, peer_port):
     global reader_added
 
     ep = Endpoint()
-    ep.connect(peer_ip, peer_port)
+    deadline = time.time() + timeout
+    connection_established = False
+    while time.time() <= deadline:
+        if -1 == ep.connect(peer_ip, peer_port):
+            await asyncio.sleep(1)
+        else:
+            connection_established = True
+            break
+
+    if not connection_established:
+        raise NameError('Timeout in connection establishment attempt')
```

Comment:
I recommend the `TimeoutError` type instead.  People know to catch those exceptions in these cases.